### PR TITLE
feat(cra_tests): #1562 CRA functional test automation: 5.1-2A-2-B: Password Rejection for M2M

### PR DESCRIPTION
**Pull request by TheSomeMan**

### CODE REVIEW

File: cra/303645/tests/test_test_5_1_2a_2_b.py [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+                    HttpHeader.WWW_AUTHENTICATE: header,
+                    HttpHeader.RUUVI_ECDH_PUBLIC_KEY: self.server_public_b64,
+                },
+                cookies={"RUUVISESSION": session_id},
+            )
+        if path == GatewayApi.AUTH and method == HttpMethod.POST:
+            if isinstance(body, dict) and body.get("login") == target.ADMIN_USERNAME:
+                session.authorized = True
+                return FakeResponse(HttpStatus.C_200_OK, {"authenticated": True})
+            return FakeResponse(HttpStatus.C_401_UNAUTHORIZED, {"authenticated": False})
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tests/test_test_5_1_2a_2_b.py:211]: Fixed

File: cra/303645/tests/test_5_1_2a_2_b.py [hunk 2]
```diff
@@ -0,0 +1,746 @@
...
+        if not condition:
+            raise InvalidSetup(f"{description} (actual: {actual!r})")
+
+    def _require_security(self, condition: bool, description: str, actual: Any = "") -> None:
+        self._record_assertion(description, condition, actual)
+        if not condition:
+            raise SecurityFailure(f"{description} (actual: {actual!r})")
+
+    def _authenticate_admin(self) -> Any:
+        result: InteractiveAuthResult = self.gateway.authenticate_interactive(ADMIN_USERNAME, self.config.gw_id)
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tests/test_5_1_2a_2_b.py]: Fixed

File: cra/303645/tests/test_5_1_2a_2_b.py [hunk 3]
```diff
@@ -0,0 +1,746 @@
...
+                HttpStatus.C_401_UNAUTHORIZED,
+                AuthMech.M2M_API_BEARER_RW,
+            ),
+        )
+        scheme: str
+        for scheme in (HttpAuthScheme.BASIC, HttpAuthScheme.DIGEST):
+            mechanism: str
+            expected: int
+            path: str
+            method: str
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tests/test_5_1_2a_2_b.py:328]: Fixed

File: cra/303645/tests/test_5_1_2a_2_b.py [hunk 4]
```diff
@@ -0,0 +1,746 @@
...
+                    result=f"PASS ({response.status_code})",
+                ),
+            )
+        current: dict[str, Any] = self._read_config(self.admin_session, "post-negative GET /ruuvi.json")
+        current_hash: str = canonical_json_hash(current)
+        self._require_security(
+            current_hash == self.prepared_hash,
+            "negative POST probes did not change configuration",
+            HashComparisonEvidence(baseline=self.prepared_hash or "", final=current_hash),
+        )
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tests/test_5_1_2a_2_b.py:428]: Fixed

File: cra/303645/tests/test_5_1_2a_2_b.py [hunk 5]
```diff
@@ -0,0 +1,746 @@
...
+            self.baseline_hash = canonical_json_hash(baseline)
+            self.evidence.write("BASELINE CONFIGURATION SHA256", self.baseline_hash)
+            self.progress("Provisioning temporary RO and RW bearer keys")
+            try:
+                self._provision()
+            except Exception:
+                self.outcomes["temporary-state setup"] = "ERROR"
+                raise
+            self.progress("Verifying the prepared temporary state")
+            self._verify_prepared()
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tests/test_5_1_2a_2_b.py]: Fixed

File: cra/303645/tasks/task_5_1_2a_2_b.md [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+
+Never convert an exception into a PASS-shaped default; there must be no silent errors.
+
+## Host-side implementation tests
+
+Follow `test_test_5_1_1_2_b.py`: `unittest`, fake sessions/responses, temporary directories, and
+injected dependencies; no live gateway. Fake-server expected Authorization values, hashes, cookies,
+and wire responses must be independent fixed fixtures or independently calculated reference vectors;
+the fake must never call the `GatewayClient` code under test to manufacture its expectations. Cover
+at least:
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tasks/task_5_1_2a_2_b.md]: This test will be added in another PR.

File: cra/303645/tests/lib/config.py [hunk 7]
```diff
@@ -1,32 +1,32 @@
...
-        / "gw_cfg_default_gen_ui.json"
+DEFAULT_GATEWAY_UI_CONFIG_PATH: Path = (
+    Path(__file__).resolve().parents[4] / "gw_cfg_default" / "gw_cfg_default_gen_ui.json"
 )
-FACTORY_RESET_MESSAGE = (
+FACTORY_RESET_MESSAGE: str = (
     "USER ACTION REQUIRED: The gateway is not in the required factory-default state. "
     "Perform a factory reset by pressing and holding the CONFIGURE button for longer "
-    "than 7 seconds, then run the test again."
+    "than 11 seconds (until the Red LED blinks every 200 ms), then run the test again."
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tests/lib/config.py]: 11 seconds is correct value. Documentation will be updated in a separate PR.

File: cra/303645/tasks/task_5_1_2a_2_b.md [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+> endpoint. Only a correct configured bearer key yields `200` from a route within its scope.
+
+Record the exact status code observed for every probe, assert against the firmware-derived matrix
+below, and never assert the idealized `401` where the firmware demonstrably returns `302`.
+
+## Firmware-derived route and status matrix
+
+Authoritative reference: [`docs/http_api.md`](../../../docs/http_api.md), sections *Authentication
+model* (113-169), *Bearer keys (RO vs RW)* (131-152), *Per-method authorization summary* (188-199),
+and the endpoint entries for `/history` (289-299) and `/ruuvi.json` (240-244, 399-412).
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tasks/task_5_1_2a_2_b.md]: Documentation will be added in a separate PR.

File: cra/303645/tests/test_5_1_2a_2_b.py [hunk 9]
```diff
@@ -0,0 +1,746 @@
...
+            )
+        current: dict[str, Any] = self._read_config(self.admin_session, "post-positive GET /ruuvi.json")
+        current_hash: str = canonical_json_hash(current)
+        self._require_security(
+            current_hash == self.prepared_hash,
+            "successful RW empty-object POST did not change configuration",
+            HashComparisonEvidence(baseline=self.prepared_hash or "", final=current_hash),
+        )
+        self.outcomes[AuthMech.M2M_API_BEARER_RO] = "PASS"
+        self.outcomes[AuthMech.M2M_API_BEARER_RW] = "PASS"
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tests/test_5_1_2a_2_b.py]: Fixed

File: cra/303645/tests/test_lib.py [hunk 10]
```diff
@@ -47,10 +50,12 @@
     API_INVENTORY,
     EXPECTED_API_INVENTORY,
     ApiRoute,
-    GatewayApi as HttpGatewayApi,
     HttpHeader,
     HttpMethod,
 )
+from lib.http_api import (
+    GatewayApi as HttpGatewayApi,
+)
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tests/test_lib.py:58]: Verified with Ruff 0.16.8 and the project configuration: the current imports pass. With combine-as-imports = false, Ruff intentionally separates aliased imports; merging these blocks produces I001. Keeping the current layout.

File: cra/303645/tests/test_5_1_2a_2_b.py [hunk 11]
```diff
@@ -0,0 +1,776 @@
...
+        self._require_setup(
+            result.challenge_response.status_code == HttpStatus.C_401_UNAUTHORIZED,
+            "GET /auth returns the interactive challenge",
+            result.challenge_response.status_code,
+        )
+        if result.auth_payload.get(GatewayCfgDesc.LAN_AUTH_TYPE) != GatewayCfgLanAuthType.DEFAULT:
+            self.factory_reset_required = True
+        self._require_setup(
+            result.auth_payload.get(GatewayCfgDesc.LAN_AUTH_TYPE) == GatewayCfgLanAuthType.DEFAULT,
+            f"interactive authentication reports {GatewayCfgLanAuthType.DEFAULT}",
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tests/test_5_1_2a_2_b.py]: Fixed

File: cra/303645/tasks/task_5_1_2a_2_b.md [hunk 12]
```diff
@@ -0,0 +1,373 @@
...
+     hyphens replaced by underscores: `test_5_1_2a_2_b.py`.
+2. Create `cra/303645/tests/test_test_5_1_2a_2_b.py` with deterministic host-side implementation
+   tests. These mock-based tests are **not** ETSI compliance evidence.
+3. Reuse the shared modules in [`cra/303645/tests/lib`](../tests/lib/). Extend `lib/` only where a
+   generally reusable, backward-compatible helper is required; do not duplicate an existing helper.
+4. Update `cra/303645/tests/requirements.txt` only if a new dependency is unavoidable. Prefer the
+   standard library and the existing `requests` / `pycryptodome` dependencies.
+
+Do not modify firmware behavior, the self-assessment source documents, the other task files, the
+shared assessment documents, or any existing test or script.
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tasks/task_5_1_2a_2_b.md]: Fixed

File: cra/303645/tasks/task_5_1_2a_2_b.md [hunk 13]
```diff
@@ -0,0 +1,373 @@
...
+Restoration is mandatory once the API keys may have been applied. In a `finally` path:
+
+1. Send `POST /ruuvi.json` with `{"lan_auth_api_key":"","lan_auth_api_key_rw":""}` using the
+   authorized `Admin` session (fallback: `Bearer <RW>`; final fallback: re-authenticate `Admin`/`gw_id`
+   — always valid because the interactive password was never changed). Log every attempt; never
+   silently suppress cleanup errors.
+2. Verify restoration: authenticate `Admin`/`gw_id`, fetch `/ruuvi.json`, require
+   `lan_auth_api_key_use is False`, `lan_auth_api_key_rw_use is False`, the canonical hash equal to the
+   step-1 baseline, and both temporary keys now rejected (`Bearer <RO>` → `302` and `Bearer <RW>` →
+   `401` on their previously-authorized routes, confirming the keys no longer authorize).
```
[the remark this answers is not in the document]
Comment (TheSomeMan) [reply cra/303645/tasks/task_5_1_2a_2_b.md]: Fixed

**Review by io53 — CHANGES_REQUESTED** [review]
This may be an edge case, but if the Admin restoration attempt fails and the session factory then raises (cra/303645/tests/test_5_1_2a_2_b.py:569), the exception escapes run()’s finally block. The outer handler returns its initial ERROR result without warning that temporary keys may still be configured. 
Lets handle session creation as a restoration failure and preserve the recovery message.
